File: shortest-palindrome.cc
// Shortest Palindrome
class Solution {
public:
  string shortestPalindrome(string s) {
    int n = s.size();
    vector<int> border(n, 0);
    int j = 0;
    for (int i = 1; i < n; i++) {
      while (j >= 0 && s[i] != s[j])
        j = j ? border[j-1] : -1;
      border[i] = ++j;
    }
    j = 0;
    for (int i = 0; i < n; i++) {
      while (j >= 0 && s[n-1-i] != s[j])
        j = j ? border[j-1] : -1;
      j++;
    }
    string ss = s.substr(j, n-j);
    reverse(ss.begin(), ss.end());
    return ss+s;
  }
};

File: shortest-palindrome.rs
impl Solution {
    pub fn shortest_palindrome(s: String) -> String {
        let a = s.as_bytes();
        let n = a.len();
        let mut border = vec![0; n];
        let mut j = 0;
        for i in 1..n {
            while j >= 0 && a[i] != a[j as usize] {
                j = if j > 0 { border[j as usize-1] } else { -1 }; }
            j += 1;
            border[i] = j;
        }
        j = 0;
        for i in 0..n {
            while j >= 0 && a[n-1-i] != a[j as usize] {
                j = if j > 0 { border[j as usize-1] } else { -1 }; }
            j += 1;
        }
        let mut ss = String::from(&s[j as usize..n]).chars().rev().collect::<String>();
        ss.push_str(&s);
        ss
    }
}
